### DISCUSSION TITLE
Alert Annotation Improvements for Dashboard Filtering and Alert History

### DISCUSSION THREAD

**Discussion by drewterry** [grafana/grafana#61713] [Alerting]
Alerts generate annotations on state change.  This is helpful in tracking alert history, as well as displaying alerts on dashboard panels.

Example annotation: 
```json
    {
        "id": 15024,
        "alertId": 24,
        "alertName": "",
        "dashboardId": 0,
        "dashboardUID": null,
        "panelId": 0,
        "userId": 0,
        "newState": "Alerting",
        "prevState": "Pending",
        "created": 1674062500497,
        "updated": 1674062500497,
        "time": 1674062500000,
        "timeEnd": 1674062500000,
        "text": "Gas Flow {alertname=Gas Flow, grafana_folder=alerts, cell=1, mode=dev} - Alerting",
        "tags": [],
        "login": "",
        "email": "",
        "avatarUrl": "",
        "data": {}
    }
```

Two improvements that would make this much more usable:
- Listing the alert labels as `tags` in the annotation
- Including alert annotation data (such as AlertValue, etc) in the annotation `data` field

*Tags*

Specifying a Dashboard UID and Panel ID in an alert rule allows the visualization of the alert on a standard dashboard using the [built-in annotation query](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/annotate-visualizations/#built-in-query).

However, most dashboards use variables to select a subset of data.  However, there is no way to filter alert annotations by these values.  Annotations support [query by tag](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/annotate-visualizations/#query-by-tag), which do allow specifying variables.  If alert labels were set in the tags field (similar to label matching in notification policy, silences, etc) annotations could be properly filtered.

As an example, in the JSON above would now be:
```json
    {
        "id": 15024,
        "alertId": 24,
        "alertName": "",
        "dashboardId": 0,
        "dashboardUID": null,
        "panelId": 0,
        "userId": 0,
        "newState": "Alerting",
        "prevState": "Pending",
        "created": 1674062500497,
        "updated": 1674062500497,
        "time": 1674062500000,
        "timeEnd": 1674062500000,
        "text": "Gas Flow {alertname=Gas Flow, grafana_folder=alerts, cell=1, mode=dev} - Alerting",
        "tags": [
          "alertname:Gas Flow", 
          "grafana_folder:alerts", 
          "cell:1",
          "mode:dev"
        ],
        "login": "",
        "email": "",
        "avatarUrl": "",
        "data": {}
    }
```

And an annotation query could filter on `cell:$cell` (assuming a variable in the chart named `cell`.

*Data*

Reviewing state changes of alerts is useful, but including additional information about what caused the state change (i.e. AlertValue) would supercharge this capability.  Fortunately, annotations should be able to support this easily by adding alert annotations (i.e. AlertValue, or custom annotations) to the `data` field for annotations.

As an example, in the JSON above would now be:
```json
    {
        "id": 15024,
        "alertId": 24,
        "alertName": "",
        "dashboardId": 0,
        "dashboardUID": null,
        "panelId": 0,
        "userId": 0,
        "newState": "Alerting",
        "prevState": "Pending",
        "created": 1674062500497,
        "updated": 1674062500497,
        "time": 1674062500000,
        "timeEnd": 1674062500000,
        "text": "Gas Flow {alertname=Gas Flow, grafana_folder=alerts, cell=1, mode=dev} - Alerting",
        "tags": [],
        "login": "",
        "email": "",
        "avatarUrl": "",
        "data": {
          "AlertValue": "100"
        }
    }
```

**Comment by MaxDiOrio** [comment]
And here is an example of why this is critical.  We have one dashboard that is filtered.  The panel has an alert created, and since the alert can be triggered for any of the filters, the panel gets littered with alert annotations instead of presenting only the alerts relevant to the filter applied.

![image](https://github.com/grafana/grafana/assets/25855524/8a163187-950f-465b-9e40-c5ba3051e0e7)

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82594. Feel free to continue the discussion around this there. Thank you!
